R/gwas_mappings.R

```diff
@@ -0,0 +1,199 @@
+#'      \item \strong{qvalue} - p-values adjusted by FDR
+#' }
+#' @export
+perform_mapping <- function(phenotype = NULL,
+                            genotype = cegwas2::snps,
+                            kinship = cegwas2::kinship,
+                            P3D = FALSE,
+                            n.PC = 0,
+                            min.MAF = 0.05,
+                            map_by_chrom = FALSE,
+                            mapping_cores = parallel::detectCores(),
+                            FDR_threshold = 0.05) {
+
+    if (any(colnames(genotype)[1:4] != c("CHROM", "POS", "REF", "ALT"))){
+        stop(message(glue::glue("The genotype matrix is not formatted correctly. Please refer to documentation")))
+    }
+
+    if (any(colnames(kinship) != rownames(kinship))){
+        stop(message(glue::glue("The kinship matrix is not formatted correctly. Please refer to documentation")))
+    }
+
+    # Clean phenotypes
+    Y = na.omit(phenotype)
+    colnames(Y) <- c("strain", "trait")
+
+    # Sort kinship matrix to to be in the same order as phenotyped strains
+    kinship_sorted = kinship[sort(row.names(kinship)),sort(colnames(kinship))]
+    # Prune kinship matrix to only include phenotyped individuals
+    K = kinship_sorted[row.names(kinship_sorted) %in% Y$strain,
+                       colnames(kinship_sorted) %in% Y$strain]
+    # Sort kinship matrix to to be in the same order as phenotyped strains
+    markers = genotype %>%
+        tidyr::unite(marker, CHROM, POS, remove = FALSE ) %>%
+        dplyr::select(marker, CHROM, POS, dplyr::everything(), -REF, -ALT)
+
+    markers_sorted <- data.frame(markers[,1:3],
+                                 markers[, sort(names(
+                                     markers[names(markers)%in%Y$strain]))])
+
+    # ID markers that have a MAF outside the user-defined range
+    keepMarkers <- data.frame(
+        MAF = apply(markers_sorted[,4:ncol(markers_sorted)],
+                    MARGIN = 1,
+                    FUN = function(x){
+                        x[x==-1] <- 0
+                        return(sum(x, na.rm = T)/length(x))}
+        )) %>%
+        dplyr::mutate(marker = markers_sorted$marker) %>%
+        dplyr::filter(MAF >= min.MAF) %>%
+        dplyr::filter(MAF <= 1 - min.MAF)
+
+    # Remove markers identified to be out of MAF range
+    M <- markers_sorted %>%
+        dplyr::filter(marker %in% keepMarkers$marker)
+
+    if (map_by_chrom) {
+        # need 6 different kinship matrices
+        # chrom2:x, map on one
+        # chrom1:5, map on x, etc
+        # then subtract blups from phenotype and map on
+        # chromosome that you are not accounting for relationship
+
+        # I think these are two ways to correct phenotype for relatedness
+        yfit <- rrBLUP::mixed.solve(y = Y$trait, K = K)
+        kfit <- rrBLUP::kin.blup(data = data.frame(Y),
+                                 geno = "strain",
+                                 pheno = "trait",
```

Comment:
this is something that i will be building up in the future. pheno=trait just defines the column to run this analysis on. 

i guess youre saying that instead of trait in Y, we should make it pheno?